### PR TITLE
Final fit after early stopping

### LINKED ISSUE #779: Final fit after early stop

### PR DESCRIPTION
**Pull request by jhfjhfj1**
### Which issue(s) does this Pull Request fix?
resolves #779 

### Pull Request Status 
In progress.

### Details of the Pull Request
If epoch is not specified, use earlystopping to do it adaptively.
If epoch is specified, but earlystopping is not in the callbacks, we use earlystopping during the search to accelerate, then we do a final fit after we found the best model to finish the rest of the epochs.
Improved some tests. Use fixed seed during unit tests.

### CODE REVIEW

File: autokeras/hypermodel/preprocessor.py [hunk 1]
```diff
@@ -406,7 +407,7 @@ class LightGBMClassifier(LightGBMModel):
 
     def __init__(self, **kwargs):
```
Comment (fchollet) [inline autokeras/hypermodel/preprocessor.py]: You may want to list the `seed` argument in this docstring above.

File: autokeras/hypermodel/preprocessor.py [hunk 2]
```diff
@@ -465,7 +466,7 @@ class LightGBMRegressor(LightGBMModel):
...
```
Comment (fchollet) [inline autokeras/hypermodel/preprocessor.py]: You may want to list the `seed` argument in this docstring above.

File: autokeras/tuner.py [hunk 3]
```diff
@@ -72,27 +68,71 @@ def load_trial(self, trial):
...
+        fit_kwargs.setdefault('epochs', None)
+
+        # Insert early-stopping permanently when epochs is None.
+        callbacks = fit_kwargs['callbacks']
+        if fit_kwargs['epochs'] is None:
+            fit_kwargs['epochs'] = 1000
+            if not any([isinstance(callback, tf.keras.callbacks.EarlyStopping)
+                        for callback in callbacks]):
+                callbacks = callbacks + [
+                    tf.keras.callbacks.EarlyStopping(patience=30)]
```
Comment (fchollet) [inline autokeras/tuner.py]: 30 seems somewhat excessive, it guarantees overfitting. ~10 seems more reasonable, no?

File: autokeras/task.py [hunk 4]
```diff
@@ -203,6 +205,11 @@ def fit(self,
             y: String, numpy.ndarray, or tensorflow.Dataset. Training data y.
                 If the data is from a csv file, it should be a string corresponding
                 to the label column.
+            epochs: Int. The number of epochs to train each model during the search.
+                If unspecified, we would use epochs equal to 1000 and early stopping
```
Comment (fchollet) [inline autokeras/task.py]: "If unspecified, by default we train for a maximum of 1000 epochs but we stop training if the validation loss stops improving for 30 epochs (unless you specified an `EarlyStopping` callback as part of the `callbacks` argument, in which case this callback will determine early stopping)."

File: autokeras/tuner.py [hunk 5]
```diff
@@ -72,27 +68,71 @@ def load_trial(self, trial):
...
+        fit_kwargs['callbacks'] = callbacks
+
+        # Insert early-stopping temporarily for the search.
```
Comment (fchollet) [inline autokeras/tuner.py]: I don't understand the logic here. `final_fit` is going to be True IFF there is no EarlyStopping callback. But the only case where there is no EarlyStopping callback is if `fit_kwargs['epochs']` is set by the user *and* the user did not pass an EarlyStopping callback.

Please clarify what the intent is, and why.

Comment (jhfjhfj1) [reply autokeras/tuner.py]: Yes, this is what we want. If the user specified the epochs and did not put EarlyStopping in the callbacks, it means the user wants the model to be trained to the specific epochs. In this case, we need the final fit to fully train the model, and will not use early stopping in the final fit.

This temporary early stopping is just for accelerating the search. It is not related to the case which user did not specify the epochs.

For the case of user not specifying epochs or user specified epochs and put earlystopping in  the callbacks, each trial would be run exactly according to the user specifications. So we don't need the final fit.

Comment (fchollet) [reply autokeras/tuner.py]: I would suggest refactoring this part of the code to make the logic easier to follow (e.g. using an enumeration of cases), and adding comments.
